ROX-29232: Validate user-provided dockerConfigs for non-UTF-8 chars

vikin91 opened the pull request:
## Description

This PR is to prevent GRPC marshaling and unmarshaling errors (leading to crash of the GRPC connection) when data modeled as `string` in the proto contain non-UTF-8 characters.

Note: this is a very limited fix, as it only validates selected types of data provided from the user. It is still possible that other not yet validated user-provided inputs may lead to grpc crash due to non-UTF-8 characters.

## Update ⚠️ 

This PR won't help with the problem. Here is what I know so far:

1. ~It is impossible to read a non-UTF-8 characters from the orchestrator (from a secret, configmap, annotation etc) and crash the grpc connection. Reading it from k8s makes it non-lethal, as most probably the non-utf-8 characters are fixed in the process of reading or writing to the secret.~
2. It is possible to crash grpc connection with a **Marshaling** error when Sensor is sending to Central only when:
    1. We use version 4.6 or earlier (due to the codec change in 4.7)
    2. We craft a broken message containing non-utf-8 in the code of Sensor
3. It is possible to crash grpc connection with a **Unmarshaling** error when Central is receiving only when:
    1. On any version
    2. We craft a broken message containing non-utf-8 in the code of Sensor.
    3. On version 4.7 the marshaling error on Sensor side does not happen.

Thus: Validation of the secrets read from k8s makes no sense and this PR will be closed.

## Update 2 ⚠️ 

Apparently storing secrets with non-uft-8 characters in k8s to make them harmless works only for some non-utf-8 characters. @dcaravel found that the following secret can lead to a crash even if stored and read from k8s:

```yaml
apiVersion: v1
data:
  .dockercfg: ewogICAgImV4YW1wbGUuY29tIjogewogICAgICAgICJhdXRoIjoibnVsbE9wN3BaUT09IgogICAgfQp9
kind: Secret
metadata:
  name: badsec
  namespace: stackrox
type: kubernetes.io/dockercfg
``` 

### User-facing documentation

- [ ] [CHANGELOG.md](https://github.com/stackrox/stackrox/blob/master/CHANGELOG.md) is updated **OR** update is not needed
- [ ] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

### Testing and quality

- [ ] the change is production ready: the change is [GA](https://github.com/stackrox/stackrox/blob/master/PR_GA.md), or otherwise the functionality is gated by a [feature flag](https://github.com/stackrox/stackrox/blob/master/pkg/features/README.md)
- [ ] CI results are [inspected](https://docs.google.com/document/d/1d5ga073jkv4CO1kAJqp8MPGpC6E1bwyrCGZ7S5wKg3w/edit?tab=t.0#heading=h.w4ercgtcg0xp)

#### Automated testing

<!--
If no tests have been contributed, please explain why unless it's obvious,
e.g., the PR is a one-line comment change.
-->

- [ ] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [ ] modified existing tests

#### How I validated my change

#### Reproduction

First, I hardcoded a dockerConfig secret (in the code) with non-UTF-8 characters*. I did something like:

```go
		registries = append(registries, &storage.ImagePullSecret_Registry{
			Name:     registryAddr + "a\xc5z",
			Username: dce.Username,
		})
```
around this place: https://github.com/stackrox/stackrox/blob/0707d4adc31045c3db24638c41ffe71180944e46/sensor/kubernetes/listener/resources/secrets.go#L315
and run that with local sensor:
```
go run tools/local-sensor/main.go -kubeconfig $HOME/.cluster1/kubeconfig -no-mem-prof -no-cpu-prof -connect-central localhost:8443
```

then in Sensor logs I observed the following error:

```
common/sensor: 2025/05/12 16:08:37.341123 central_receiver_impl.go:57: Info: Stopping with error: rpc error: code = Internal desc = grpc: failed to unmarshal the received message: string field contains invalid UTF-8
common/sensor: 2025/05/12 16:08:37.341145 central_sender_impl.go:116: Error: unable to send to stream: EOF
common/sensor: 2025/05/12 16:08:37.341152 central_communication_impl.go:73: Error: Stopping connection due to error: rpc error: code = Internal desc = grpc: failed to unmarshal the received message: string field contains invalid UTF-8
```

*) I did that, because I am unable to type a non-UTF-8 character from my keyboard directly into a k8s secret.

dcaravel commented on sensor/kubernetes/listener/resources/secrets.go:
```diff
@@ -290,6 +307,11 @@ func (s *secretDispatcher) processDockerConfigEvent(secret, oldSecret *v1.Secret
 		}
 
+		if err := validateSecret(registryAddr, dce, secret.GetNamespace(), secret.GetName()); err != nil {
+			log.Warnf("Skipping dockerConfig secret: %v", err)
```
Be good log the namespace + secret name as well (so that can locate the bad data)

vikin91 replied:
The `err` contains details about the name of secret and namespace. This line just wraps that error and prints it, so the information will be there.

dcaravel reviewed (commented):
Few top of mind thoughts
- This seems like it would work just fine, bit shocked we haven't run into this in the past.
- Is the issue only with gRPC such that k8s / CRI-O still work OK even when a secret has a 'non-utf-8' character in it?
  - If so, wondering if worth looking at how k8s/cri-o handle this - perhaps mimcing any encoding / types / validation would be better? (granted introduces some other complexes on either end of the comms, and when persisting the data to Central DB may run into the same error during marshalling)
  - Or perhaps there is some gRPC flag to not be strict on invalid utf-8 chars? <- Not so sure I like this - is a cover up
  - Or perhaps `strings.ToValidUTF8` can help? <- this also makes me uneasy as the data is being changed.
 

Would it help to also add logging to the error message on either end of the comms to print the 'type' information of the message that failed to be marshalled/unmarshalled? (ie: error failed to marshall, bad utf8, in MsgFromSensor.SensorEvent_ImageIntegration)  so that we would know (for example) it is a secret that had the bad data vs. a deployment vs. something else.

vikin91 commented:
>Would it help to also add logging to the error message on either end of the comms to print the 'type' information of the message that failed to be marshalled/unmarshalled? (ie: error failed to marshall, bad utf8, in MsgFromSensor.SensorEvent_ImageIntegration) so that we would know (for example) it is a secret that had the bad data vs. a deployment vs. something else.

It is a second case like this that I am handling and I would love to have the information that you suggest. The tricky part here is that if there is an error returned from grpc (on receiving), then the `msg` object is nil. In order to get to that object, we would need to change the automatically-generated grpc code. So shortly speaking, it is impossible in current case.

However, it is possible to do that on the sending side. And I think, we should have it.

janisz commented:
I made a small test and looks like VT proto properly handles non-utf8 characters. From your example it looks proto is called instead of vt proto. This may happen because codec is not setup properly (in theory it should as it's in `init` and sensor imports `pkg/grpc`). I curious why this was not happening before as gogo also reports an error on non-utf8 character. 

- https://github.com/stackrox/stackrox/pull/14083

```go
package protocompat

import (
	"testing"

	gogo "github.com/gogo/protobuf/proto"
	"github.com/stackrox/rox/pkg/protoassert"
	"github.com/stretchr/testify/assert"
	"google.golang.org/protobuf/proto"

	"github.com/stackrox/rox/generated/storage"
)

var name = string([]byte{0xC0})

func TestUnmarshalGogo(t *testing.T) {
	m := &storage.NamespaceMetadata{Name: name}
	_, err := gogo.Marshal(m)
	assert.EqualError(t, err, `proto: field ".Name" contains invalid UTF-8`)
}

func TestUnmarshalVT(t *testing.T) {
	m := &storage.NamespaceMetadata{Name: name}
	a, err := m.MarshalVT()
	assert.NoError(t, err)
	n := &storage.NamespaceMetadata{}
	err = n.UnmarshalVT(a)
	assert.NoError(t, err)
	protoassert.Equal(t, m, n)
}

func TestUnmarshalProto(t *testing.T) {
	m := &storage.NamespaceMetadata{Name: name}
	_, err := proto.Marshal(m)
	assert.EqualError(t, err, `string field contains invalid UTF-8`)
}
```

dcaravel commented:
> It is a second case like this that I am handling and I would love to have the information that you suggest. The tricky part here is that if there is an error returned from grpc (on receiving), then the `msg` object is nil. In order to get to that object, we would need to change the automatically-generated grpc code. So shortly speaking, it is impossible in current case.
> 
> However, it is possible to do that on the sending side. And I think, we should have it.

Agree not possible on the receiving side, would be on the sending side.

janisz commented:
We can create a unit test for straming like here https://github.com/stackrox/stackrox/pull/14083/files#diff-80aadd25196a5db848dfb2c5d9240d09512010efc442f90766ac247fbcfc4180 
Or you can change the error log line to "+v" to get stack trace if error is wrapped

vikin91 commented:
Temporary, one can use branch `piotr/ROX-29232-broken-docker-secrets` to make sensor send broken messages to Central with command:

```
ROX_LOCAL_IMAGE_SCANNING_ENABLED=true go run tools/local-sensor/main.go -kubeconfig $HOME/.cluster1/kubeconfig -no-mem-prof -no-cpu-prof -connect-central localhost:8443
```

vikin91 commented:
> #### ⚠️ GitGuardian has uncovered 1 secret following the scan of your pull request.
> Please consider investigating the findings and remediating the incidents. Failure to do so may lead to compromising the associated services or software components.

This is a dummy secret created only for the tests.

RTann reviewed (commented):
Out of curiosity, do you know what happens in K8s? As in, does K8s see these registries as valid/are they actually valid?

vikin91 commented:
> Out of curiosity, do you know what happens in K8s? As in, does K8s see these registries as valid/are they actually valid?

Unfortunately, I don't know. I would need to do an experiment.

dcaravel reviewed (commented):
Is it worth preventing the connection from being killed for any invalid message? Will that be too much overhead? Since this can occur for any message sent to Central from Sensor (or vice versa).

I realize that k8s does some degree of validation for more structured fields (such as resource names) - should we treat all info from k8s events as 'untrusted input'?

vikin91 commented on sensor/kubernetes/listener/resources/secrets.go:
```diff
@@ -290,6 +307,11 @@ func (s *secretDispatcher) processDockerConfigEvent(secret, oldSecret *v1.Secret
 				registryAddress, secret.GetNamespace(), secret.GetName())
 		}
 
+		if err := validateSecret(registryAddr, dce, secret.GetNamespace(), secret.GetName()); err != nil {
+			log.Warnf("Not adding registry %s from dockerConfig secret: %v", registryAddr, err)
+			continue
+		}
+
```
> What are your thoughts on moving the validity check higher, so that we also do not attempt to use the invalid secret for delegated scanning?

It is our decision, but one secret can contain auth data for multiple registries and only one of those registries may have the auth string broken. It would be pity to reject all other entries from that secret, or do you see it differently?

vikin91 replied:
> I am assuming if this issue is caused by a limitation of our chosen gRPC libraries / configuration it will affect comms from Sensor -> Scanner V4 Indexer as well.

This may be it, but not fully necessary. The secret may be simply broken. Imagine someone copy-pasted the base64-encoded auth part of the secret and missed single character when marking the text to be copied. Then by decoding that from base64 we will always get a broken auth data. K8s could fix that by changing some non-utf-8 chars to some similar ones, but in my opinion, the auth would not work. 
So if our grpc codec would magically fix the utf-8 issues, the auth data would be not usable.

I think, that in such situations, we should raise the alarm that something is wrong as soon as possible - in our case this would be in Sensor.

I still think, that all our grpc connections should be robust against this kind of problems, but I believe that we should implement both: input validation and grpc codec robustness, not one xor the other.

vikin91 replied:
I was thinking that the path from Sensor to Local Scanner could be affected IF the `ImageIntegrations` message is exchanged over grpc between them (or any other message that includes the decoded and parsed data that is stored as base64 in k8s).

vikin91 commented on sensor/kubernetes/listener/resources/secrets.go:
```diff
@@ -260,6 +261,22 @@ func shouldCreateSourcedIntegration(secret *v1.Secret) bool {
 		(env.AutogenerateGlobalPullSecRegistries.BooleanSetting() && openshift.GlobalPullSecret(secret.GetNamespace(), secret.GetName()))
 }
 
+func validateSecret(registryAddr string, dce config.DockerConfigEntry, namespace, secretName string) error {
+	if !utf8.ValidString(registryAddr) {
+		return fmt.Errorf("registry address %q contains invalid UTF-8 characters. "+
+			"Correct the contents of secret %s/%s", registryAddr, namespace, secretName)
+	}
+	if !utf8.ValidString(dce.Username) {
+		return fmt.Errorf("registry username %q contains invalid UTF-8 characters. "+
+			"Correct the contents of secret %s/%s", registryAddr, namespace, secretName)
+	}
+	if !utf8.ValidString(dce.Password) {
+		return fmt.Errorf("registry password located in secret %s/%s contains invalid UTF-8 characters",
+			namespace, secretName)
+	}
+	return nil
+}
+
```
First bullet: Yes, the call to action (using the word "correct") is missing. I can add it. I chose different wording to not print the password to the console.

vikin91 replied:
Addressed (partially) here: [9da1da2](https://github.com/stackrox/stackrox/pull/15270/commits/9da1da2bf5a0d3287f1942f0566a5556f2b658e1)

vikin91 commented on sensor/kubernetes/listener/resources/secrets.go:
```diff
@@ -260,6 +261,22 @@ func shouldCreateSourcedIntegration(secret *v1.Secret) bool {
 		(env.AutogenerateGlobalPullSecRegistries.BooleanSetting() && openshift.GlobalPullSecret(secret.GetNamespace(), secret.GetName()))
 }
 
+func validateSecret(registryAddr string, dce config.DockerConfigEntry, namespace, secretName string) error {
+	if !utf8.ValidString(registryAddr) {
+		return fmt.Errorf("registry address %q contains invalid UTF-8 characters. "+
+			"Correct the contents of secret %s/%s", registryAddr, namespace, secretName)
+	}
+	if !utf8.ValidString(dce.Username) {
+		return fmt.Errorf("registry username %q contains invalid UTF-8 characters. "+
+			"Correct the contents of secret %s/%s", registryAddr, namespace, secretName)
+	}
+	if !utf8.ValidString(dce.Password) {
+		return fmt.Errorf("registry password located in secret %s/%s contains invalid UTF-8 characters",
+			namespace, secretName)
+	}
+	return nil
+}
```
I am returning here with a second thoughts: it looks like we are not generating a generic `SensorEvent_Secret` if the secret is of types `v1.SecretTypeDockerConfigJson, v1.SecretTypeDockercfg` and it contains no valid registry entries. 
In that case, I change my mind and will not emit that generic event if the entries in the secret are failing the validation. Is that okay?

vikin91 commented:
@dcaravel I addressed some of the issues; others I commented on as I am not sure how to incorporate them into this PR. Do you think we can merge this for 4.8 or shall we improve more?

janisz commented on sensor/kubernetes/listener/resources/secrets.go:
```diff
@@ -290,6 +304,13 @@ func (s *secretDispatcher) processDockerConfigEvent(secret, oldSecret *v1.Secret
 				registryAddress, secret.GetNamespace(), secret.GetName())
 		}
 
+		if err := validateSecret(registryAddr, dce); err != nil {
```
Should we validate configs before using them so right after `getDockerConfigFromSecret` remove invalid configs and then check the len.

vikin91 replied:
Yes, the second idea is something that would work and makes sense.
Validation right after `getDockerConfigFromSecret` would require to iterate over the registries only to do the validation, so it would be similar to current approach.

I will give it a try to validate in `config.go`.

vikin91 replied:
However, in `func decodeDockerConfigFieldAuth(field string) (username, password string, err error) {`, we only decode username, password and auth string. The registry address is missing. We would need to validate in both places to cover everything (which is maybe not so bad).

vikin91 replied:
I tried it and there is one more problem with that approach: because of this https://github.com/stackrox/stackrox/blob/9da1da2bf5a0d3287f1942f0566a5556f2b658e1/sensor/kubernetes/listener/resources/secrets.go#L225-L228, we would reject the entire secret if at least one registry contains invalid data. If that secret contains more registries, we would loose them. WDYT?

janisz replied:
> The registry address is missing. 

Is it possible to have non UTF-8 character in address? Maybe this should be validated by k8s I reported it.
 - https://github.com/kubernetes/kubernetes/issues/131982
 
 
 > we would reject the entire secret if at least one registry contains invalid data. If that secret contains more registries, we would loose them. WDYT?
 
 I we can inform User about it I think it will be fine but with current implementation it's only logged in sensor and this error will not be passed to central.

vikin91 replied:
It should not be possible, but I cannot exclude some exotic scenarios, where base64 encoded secrets are copy-pasted manually and some data may get corrupted unintentionally. I find checking that really cheap, so maybe we can keep it?

janisz reviewed (commented):
I think we need a changelog entry for this with information that we do not support non UTF-8 docker configs.

rhybrillou commented on sensor/kubernetes/listener/resources/secrets.go:
```diff
@@ -290,6 +304,13 @@ func (s *secretDispatcher) processDockerConfigEvent(secret, oldSecret *v1.Secret
 
+		if err := validateSecret(registryAddr, dce); err != nil {
+			log.Warnf("Not processing the registry with address %q found in secret %q from namespace %q: %v. ",
+				secret.GetName(), secret.GetNamespace(), registryAddr, err)
```
Would it make sense to have information on skipped items in the administative events ?
This would be follow-up work

rhybrillou replied:
Good point. I don't think so.

vikin91 commented on sensor/common/sensor/central_sender_impl.go:
```diff
@@ -1,8 +1,7 @@
 import (
-	"errors"
-
+	"github.com/pkg/errors"
```
Apparently, I did A->B->A and some automated changes from B stayed. I can drop it.

vikin91 replied:
Done in [d4453b1](https://github.com/stackrox/stackrox/pull/15270/commits/d4453b164d5b373185e279d65d68f75b37354e28)

vikin91 commented on sensor/kubernetes/listener/resources/secrets.go:
```diff
@@ -290,6 +304,13 @@ func (s *secretDispatcher) processDockerConfigEvent(secret, oldSecret *v1.Secret
 		}
 
+		if err := validateSecret(registryAddr, dce); err != nil {
+			log.Warnf("Not processing the registry with address %q found in secret %q from namespace %q: %v. ",
```
Done in [d4453b1](https://github.com/stackrox/stackrox/pull/15270/commits/d4453b164d5b373185e279d65d68f75b37354e28)

vikin91 commented:
@dcaravel I acknowledge that this PR is not fixing all the issues, but I believe that having it in 4.8 would be valuable.
Thus, I plan to merge it, unless you give it a :x:. Would that be okay?

vikin91 commented:
The integration tests failure looks unrelated:

```
    sync_test.go:100: 
        	Error Trace:	/go/src/github.com/stackrox/stackrox/sensor/tests/complianceoperator/sync_test.go:100
        	            				/go/src/github.com/stackrox/stackrox/sensor/tests/helper/helper.go:396
        	            				/go/src/github.com/stackrox/stackrox/sensor/tests/helper/helper.go:352
        	            				/go/src/github.com/stackrox/stackrox/sensor/tests/helper/helper.go:231
        	            				/go/src/github.com/stackrox/stackrox/sensor/tests/complianceoperator/sync_test.go:89
        	Error:      	Received unexpected error:
        	            	object is being deleted: customresourcedefinitions.apiextensions.k8s.io "scansettings.compliance.openshift.io" already exists
        	Test:       	Test_ComplianceOperatorScanConfigSync
```
